src/storage/sql_table.cpp

```diff
@@ -19,64 +43,329 @@ SqlTable::SqlTable(BlockStore *const store, const catalog::Schema &schema, const
+    auto col_oid = column.GetOid();
+    auto *default_value = column.GetDefault();
+    // Only populate the default values of the columns which are new and have a default value
+    if (default_value_map_.Find(col_oid) == default_value_map_.End()) {
+      uint8_t attr_size = column.GetAttrSize();
+      byte *temp = nullptr;
+      if (default_value != nullptr) {
+        // clang's memory analyis has a false positive on this allocation.  All memory
+        // allocated here is placed into 'default_value_map_' and eventually freed
+        // with the rest of the SqlTable object in the destructor (see line 20).
+        // NOLINTNEXTLINE
+        temp = new byte[attr_size];
+        std::memcpy(temp, default_value, attr_size);
+      }
+      default_value_map_.Insert(column.GetOid(), {temp, attr_size});
+    }
+  }
+
+  auto layout = BlockLayout(attr_sizes);
+
+  auto dt = new DataTable(block_store_, layout, schema.GetVersion());
+  // clang's memory analysis has a false positive on this allocation.  The TERRIER_ASSERT on the second line of this
+  // function prevents the insert below from failing (can only fail when key is not unique).  The write-lock on the
+  // catalog prevents any other transaction from being in a race condition with this one.  The corresponding delete
+  // for this allocation is in the destructor for SqlTable.  clang-analyzer-cplusplus.NewDeleteLeaks identifies this
+  // as a potential leak and throws an error incorrectly.
+  // NOLINTNEXTLINE
+  tables_.Insert(schema.GetVersion(), {dt, layout, col_map, inv_col_map});
 }
 
-std::vector<col_id_t> SqlTable::ColIdsForOids(const std::vector<catalog::col_oid_t> &col_oids) const {
+bool SqlTable::Select(transaction::TransactionContext *const txn, const TupleSlot slot, ProjectedRow *const out_buffer,
+                      const ProjectionMap &pr_map, layout_version_t version_num) const {
+  TERRIER_ASSERT(slot.GetBlock() != nullptr, "Slot does not exist");
+  STORAGE_LOG_DEBUG("slot version: {}, current version: {}", !slot.GetBlock()->layout_version_, !version_num);
+
+  layout_version_t old_version_num = slot.GetBlock()->layout_version_;
+
+  TERRIER_ASSERT(out_buffer->NumColumns() <= tables_.Find(version_num)->second.column_map.size(),
+                 "The output buffer never returns the version pointer columns, so it should have "
+                 "fewer attributes.");
+
+  // The version of the current slot is the same as the version num
+  if (old_version_num == version_num) {
+    return tables_.Find(version_num)->second.data_table->Select(txn, slot, out_buffer);
+  }
+
+  auto old_dt_version = tables_.Find(old_version_num)->second;
+  auto curr_dt_version = tables_.Find(version_num)->second;
+
+  // The slot version is not the same as the version_num
+  col_id_t original_column_ids[out_buffer->NumColumns()];
+  ModifyProjectionHeaderForVersion(out_buffer, tables_.Find(version_num)->second, old_dt_version, original_column_ids);
```

Comment:
tables_.Find(version_num)->second is repeated several times here, and curr_dt_version is unused. Are you doing this for some concurrent issues?